### PR TITLE
Asynchronous connection with example

### PR DESCRIPTION
**Pull request by FrancMunoz**
Changed some things in GSM and GPRS classes to be able to manage an asynchronous connection. A working example is provided too.

### CODE REVIEW

File: src/GPRS.h [hunk 1]
```diff
@@ -89,6 +112,7 @@ class GPRS : public ModemUrcHandler {
   int ping(IPAddress ip, uint8_t ttl = 128);
 
   GSM3_NetworkStatus_t status();
+  int getReadyState();
```
Comment (sandeepmistry) [inline src/GPRS.h]: I'm not too keen on exposing the ready state internals. Is this needed?

File: src/GSM.h [hunk 2]
```diff
@@ -71,6 +91,9 @@ class GSM {
   int lowPowerMode();
   int noLowPowerMode();
 
+  GSM3_NetworkStatus_t status();
+  int getReadyState();
```
Comment (sandeepmistry) [inline src/GSM.h]: I'm not too keen on exposing the ready state internals. Is this needed?

File: src/GSM.h [hunk 3]
```diff
@@ -71,6 +91,9 @@ class GSM {
   int lowPowerMode();
   int noLowPowerMode();
 
+  GSM3_NetworkStatus_t status();
```
Comment (sandeepmistry) [inline src/GSM.h]: Adding API for the status is a good idea!

File: examples/GsmAsyncConnection/GsmAsyncConnection.ino [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      case CONNECTING: // connecting...
+        if(lastState!=state) {
+          //Serial.print("State changed: ");
+          //Serial.println(state);
+          lastState=state;
+        }
+        gsm.ready();
+        break;
+      case GSM_READY: // connected, let's wait to READY_STATE_DONE
+        if(gsm.getReadyState() == READY_STATE_DONE) {
```
Comment (sandeepmistry) [inline examples/GsmAsyncConnection/GsmAsyncConnection.ino]: Can we just use `gsm.ready()` API here?

File: examples/GsmAsyncConnection/GsmAsyncConnection.ino [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        case CONNECTING: // connecting...
+          if(lastState!=state) {
+            //Serial.print("State changed: ");
+            //Serial.println(state);
+            lastState=state;
+          }
+          gprs.ready();
+          break;
+        case GPRS_READY: // connected, let's wait to GPRS_STATE_IDLE
+          if(gprs.getReadyState() == GPRS_STATE_IDLE) {
```
Comment (sandeepmistry) [inline examples/GsmAsyncConnection/GsmAsyncConnection.ino]: Can we just using `gprs.read()` here?

File: src/GPRS.h (hunk 1 above)
Comment (FrancMunoz) [reply src/GPRS.h]: Not, is not really needed... I used to know the progress of the connection while debugging. Also, I've noticed using GSM::ready() returns 1 only when the progress has finished... so It should be removed.

File: examples/GsmAsyncConnection/GsmAsyncConnection.ino (hunk 4 above)
Comment (FrancMunoz) [reply examples/GsmAsyncConnection/GsmAsyncConnection.ino]: Sure, it's the best way... I complicated my life... :)

File: examples/GsmAsyncConnection/GsmAsyncConnection.ino (hunk 5 above)
Comment (FrancMunoz) [reply examples/GsmAsyncConnection/GsmAsyncConnection.ino]: Yes, the same for GSM...

**Comment by FrancMunoz** [comment]
Hi @sandeepmistry ,

Reviewing #44 , I've tested without exposing readystate (I must admit I didn't see i was complicating my life) and as I guessed works perfectly... 

I'd change PR when #44 it's merged so the example will use new buffered read.

What do you think?

**Comment by sandeepmistry** [comment]
@FrancMunoz let's split out the new API you proposed and example into two pull requests.

If you could open a new one for the `GSM::status()` API to get things rolling that would be great.

**Comment by sandeepmistry** [comment]
I've added `GSM:status()` in PR #53. @FrancMunoz your feedback is appreciated as always.

**Comment by Rocketct** [comment]
Closing in favour of https://github.com/arduino-libraries/MKRGSM/pull/53

**Pull request closed without merging**